### PR TITLE
Jiayu Liu Task2&4

### PR DESCRIPTION
**Pull request by jiayu-Liu0407**
不需要merge
My tank:LjyTank
My issue:
https://github.com/Yao-class-cpp-studio/battle_game/issues/178 LaserTank:Add a new skill "Destroy",which can use by press 'E' while left botton of mouse not being presssed.
#379 CollisionTank:Build the basis of collision tank,concluding damage and move system,when I collise to another tank,two tank will suffer 5 and 20 damage (after Zkx improved moving system,I change the damage to 5*speed and 20*speed,the tank faster,the damage higher)

### CODE REVIEW

**Comment by jiayu-Liu0407** [comment]
#379 Collision Tank:modify the damage system,damage now is 5*speed and 20*speed.when it is faster,damage will increase.

**Review by jiayu-Liu0407 — COMMENTED** [review]
#178 Add bullet laser

File: src/battle_game/core/bullets/laser.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+ private:
+  void update_color(float &a);
+  float angle();
+  float Angle_();        // ComboTime, LightIndex_
+  float Transparent_();  // ComboTime
+  float Width_();        // ComboTime
+  float White_(int Distance);
+  int ComboTime;
+  int LightIndex_;
```
Comment (qinchuanhui) [inline src/battle_game/core/bullets/laser.h:29]: Function naming: `UpdateColor`, `GetWidth` (and I don't know what does 'width' mean, similar to other function names)
Private variables naming: `light_index_`, `combo_time_`

File: src/battle_game/core/units/collision_tank.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+      /* Tank Turret */
+      std::vector<ObjectVertex> turret_vertices;
+      std::vector<uint32_t> turret_indices;
+      const int precision = 60;
+      const float inv_precision = 1.0f / float(precision);
+      for (int i = 0; i < precision; i++) {
+        auto theta = (float(i) + 0.5f) * inv_precision;
+        theta *= glm::pi<float>() * 2.0f;
+        auto sin_theta = std::sin(theta);
+        auto cos_theta = std::cos(theta);
```
Comment (qinchuanhui) [inline src/battle_game/core/units/collision_tank.cpp:47]: Inherit from `Tank` class if u don't change these functions

File: src/battle_game/core/units/laser_tank.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        auto sin_theta = std::sin(theta);
+        auto cos_theta = std::cos(theta);
+        turret_vertices.push_back({{sin_theta * 0.5f, cos_theta * 0.5f},
+                                   {0.0f, 0.0f},
+                                   {0.7f, 0.7f, 0.7f, 1.0f}});
+        turret_indices.push_back(i);
+        turret_indices.push_back((i + 1) % precision);
+        turret_indices.push_back(precision);
+      }
+      turret_vertices.push_back(
```
Comment (qinchuanhui) [inline src/battle_game/core/units/laser_tank.cpp:120]: The same, just inherit `Tank` class

File: src/battle_game/core/units/ljy_tank.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    {
+      /* Tank Turret */
+      std::vector<ObjectVertex> turret_vertices;
+      std::vector<uint32_t> turret_indices;
+      const int precision = 60;
+      const float inv_precision = 1.0f / float(precision);
+      for (int i = 0; i < precision; i++) {
+        auto theta = (float(i) + 0.5f) * inv_precision;
+        theta *= glm::pi<float>() * 2.0f;
```
Comment (qinchuanhui) [inline src/battle_game/core/units/ljy_tank.cpp:41]: The same issue.

File: src/battle_game/core/units/laser_tank.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  void Fire();
+  [[nodiscard]] const char *UnitName() const override;
+  [[nodiscard]] const char *Author() const override;
+
+  float turret_rotation_{0.0f};
+  uint32_t fire_count_down_{0};
+  bool iscombo{false};
+  bool is_stuck_by_swamp{false};
+  uint32_t destroy_count_down_{600};
+  uint32_t isDestroy = 0;
```
Comment (qinchuanhui) [inline src/battle_game/core/units/laser_tank.h:26]: Naming issue

File: src/battle_game/core/units/collision_tank.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  void AlreadyHit();
+  void TurretRotate();
+  void Fire();
+  [[nodiscard]] const char *UnitName() const override;
+  [[nodiscard]] const char *Author() const override;
+
+  float turret_rotation_{0.0f};
+  uint32_t fire_count_down_{0};
+  uint32_t mine_count_down_{0};
+  std::map<uint32_t, bool> alreadyhit;
```
Comment (qinchuanhui) [inline src/battle_game/core/units/collision_tank.h:24]: Naming issue

**Review by qinchuanhui — CHANGES_REQUESTED** [review]

**Review by qinchuanhui — APPROVED** [review]
OK, u can get the corresponding score

**Pull request closed without merging**